src/tree_annotation/database.cljs

```diff
@@ -25,172 +27,323 @@
 (defn set-input-str [input-str]
   (swap! db assoc :input-str input-str))
 
-;-----------------------;
-; Input string requests ;
-;-----------------------;
+;----------------------------;
+; Input tree string requests ;
+;----------------------------;
 
 (defn get-input-tree-str []
   (:input-tree-str @db))
 
 (defn set-input-tree-str [input-tree-str]
   (swap! db assoc :input-tree-str input-tree-str))
 
+(defn strip-math? []
+  (:strip-math @db))
+
+(defn toggle-strip-math! []
+  (swap! db update :strip-math not))
+
 ;------------------------;
 ; Output string requests ;
 ;------------------------;
 
-(defn get-output-str []
-  (:output-str @db))
-
-(defn set-output-str [output-str]
-  (swap! db assoc :output-str output-str))
-
-;-----------------------;
-; Rename label requests ;
-;-----------------------;
-
-(defn get-rename-label []
-  (:rename-label @db))
-
-(defn set-rename-label [label]
-  (swap! db assoc :rename-label label))
+(declare leaf?)
+
+(defn tree-str [math-inner math-leaves node]
+  "Converts `node` into a qtree string.
+`math-inner` and `math-leaves` are booleans that indicate whether labels
+of inner and leaf nodes should be enclosed $s, respecively."
+  (let [children (:children node)
+        label    (:label node)
+        math     (or (and (leaf? node) math-leaves)
+                     (and (not (leaf? node)) math-inner))
+        wrap     (fn [s] (if math (str "$" s "$") s))]
+    (if (empty? children) ; if the node is a leaf
+      (str (wrap label) " ")
+      (apply str
+             (concat ["[." (wrap label) " "]
+                     (map (partial tree-str math-inner math-leaves) children)
+                     ["] "])))))
 
-(defn get-node-coords-under-renaming []
-  (->> (:nodes @db)
-       (filter (fn [[coord {state :state}]] (= :rename state)))
-       (map key)))
-
-;---------------------;
-; Node state requests ;
-;---------------------;
-
-(defn get-node-state [coord] 
-  (get-in @db [:nodes coord :state]))
-
-(defn set-node-state [coord new-state]
-  (swap! db assoc-in [:nodes coord :state] new-state))
-
-(defn toggle-select [coord]
-  "Select a node if it is not selected or 
-   unselect it if it is selected."
-  (let [new-state (case (get-node-state coord)
-                     :selected :not-selected
-                     :not-selected :selected)]
-    (set-node-state coord new-state)))
-
-;---------------------;
-; Node label requests ;
-;---------------------;
-
-(defn get-node-label [coord] 
-  (get-in @db [:nodes coord :label]))
+(defn get-output-str []
+  "Returns the qtree string representation of the forest."
+  (let [math-inner (:math-inner @db)
+        math-leaves (:math-leaves @db)]
+    (apply str (map (partial tree-str math-inner math-leaves)
+                    (:forest @db)))))
```

Comment:
It makes sense to use these functions in the let binding. However, the reason for binding them is that they are used in a closure that is mapped over the forest, so it makes sense to query the values once instead of querying them every time the closure is run.

Reply:
However, the best solution is probably to dereference `db` just once, instead of doing separately for `math-inner`, `math-leaves`, and `forest`.

Reply:
In that case the query functions should be reserved for when the individual information should be dereferenced (such as for rendering the checkboxes).